### PR TITLE
8270366: C2: Add associative rule to add/sub node

### PR DESCRIPTION
**Pull request by zhengyu123**
Please review this small patch that add associative rule to add/sub node, that eliminates a multiply instruction.

e.g.
```
private static int assocInt(int a, int b, int c) {
  return a * c + b * c;
}

```
x86_64:
Before:
```
  0x00007fda1506152c:   imul   %ecx,%esi
  0x00007fda1506152f:   imul   %ecx,%edx
  0x00007fda15061532:   mov    %esi,%eax
  0x00007fda15061534:   add    %edx,%eax                    ;*iadd {reexecute=0 rethrow=0 return_oop=0}
                                                            ; - TestAssoc::assocInt@6 (line 9)

```
After:
```
  0x00007fc1c078d52c:   add    %edx,%esi
  0x00007fc1c078d52e:   mov    %ecx,%eax
  0x00007fc1c078d530:   imul   %esi,%eax                    ;*iadd {reexecute=0 rethrow=0 return_oop=0}
                                                            ; - TestAssoc::assocInt@6 (line 9)
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8270366](https://bugs.openjdk.java.net/browse/JDK-8270366): C2: Add associative rule to add/sub node

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Andrew Dinn](https://openjdk.java.net/census#adinn) (@adinn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4765/head:pull/4765` \
`$ git checkout pull/4765`

Update a local copy of the PR: \
`$ git checkout pull/4765` \
`$ git pull https://git.openjdk.java.net/jdk pull/4765/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4765`

View PR using the GUI difftool: \
`$ git pr show -t 4765`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4765.diff">https://git.openjdk.java.net/jdk/pull/4765.diff</a>

</details>

### CODE REVIEW

**Comment by adinn** [comment]
Hi Zhengyu,

It looks like you are missing a case for both long and int. Your code looks for these patterns:

    a*b+a*c --> a*(b+c)
    a*b+b*c --> b*(a+c)
    a*c+b*c --> (a+b)*c

So you are not transforming this this case

  a*b+c*a -> a*(b+c)

**Comment by dean-long** [comment]
Can this rule cause the JIT to compute different results than the interpreter if there is integer overflow?  Does the spec allow that?

**Comment by vnkozlov** [comment]
@jddarcy may know the answer.

**Comment by adinn** [comment]
> Can this rule cause the JIT to compute different results than the interpreter if there is integer overflow?

I don't believe so.

It should obvious that unsigned arithmetic will always give the same result for any order of computation. Why? Well, if we compute the result for a sequence of n-bit operations in an unlimited bit field we can be sure the results are order independent because of the associativity law. Setting and clearing of higher bits because of overflow beyond n-bits will not change the result that appears in the low n bits.

For signed arithmetic all we need to note is that for every signed n-bit operation that overflows (or underflows) there is an equivalent unsigned operation in 2*n bits that has the same bottom n-bits plus some extra non-zero bits in the top n bits. So, for any chain of k signed operations there will be an equivalent chain of unsigned operations in at most k*n bits which has the same low n bits. Re-ordering the unsigned operations would produce the same unsigned result which guarantees that an equivalent re-ordering of the signed operations will produce the same signed value in the low n bits.

**Comment by zhengyu123** [comment]
> Hi Zhengyu,
> 
> It looks like you are missing a case for both long and int. Your code looks for these patterns:
> 
> ```
> a*b+a*c --> a*(b+c)
> a*b+b*c --> b*(a+c)
> a*c+b*c --> (a+b)*c
> ```
> 
> So you are not transforming this this case
> 
> ```
> a*b+c*a -> a*(b+c)
> ```

**Comment by zhengyu123** [comment]
> Hi Zhengyu,
> 
> It looks like you are missing a case for both long and int. Your code looks for these patterns:
> 
> ```
> a*b+a*c --> a*(b+c)
> a*b+b*c --> b*(a+c)
> a*c+b*c --> (a+b)*c
> ```
> 
> So you are not transforming this this case
> 
> ```
> a*b+c*a -> a*(b+c)
> ```


Thanks, Andrew.

Updated.

**Comment by zhengyu123** [comment]
> Can this rule cause the JIT to compute different results than the interpreter if there is integer overflow? Does the spec allow that?

Based on [https://www.agner.org/optimize/optimizing_cpp.pdf](Optimizing software in C++), this optimization is implemented in all 4 compilers (GNU, Clang, Microsoft and Intel) author investigated.

For example:
```
long mask(long a, long b, long c) {
    return a*b + a*c ;
}
```
GNU/Clang:

```
 lea    rax,[rdx+rsi*1]
 imul   rax,rdi
```
MSVC:   

```
lea     eax, DWORD PTR [rdx+r8]
imul    eax, ecx

```

**Comment by vnkozlov** [comment]
https://docs.oracle.com/javase/specs/jls/se16/html/jls-15.html#jls-15.17.1
```
If an integer multiplication overflows, then the result is the low-order bits of the mathematical product as represented in some sufficiently large two's-complement format. As a result, if overflow occurs, then the sign of the result may not be the same as the sign of the mathematical product of the two operand values.
```

I am interpreting this as using long arithmetic for integer overflow and taking low 32-bits of result. In this sense suggested optimization is legal.

*But* we may have bugs in VM. I suggest to add a regression test to compare golden (from Interpreter) values with results from optimized compiled code for all interesting/corner cases.

**Comment by chhagedorn** [comment]
This RFE would also be well suited for using the new [IR framework](https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/lib/ir_framework/README.md) to create an IR test to verify the new IR transformations.

**Comment by zhengyu123** [comment]
> https://docs.oracle.com/javase/specs/jls/se16/html/jls-15.html#jls-15.17.1
> 
> ```
> If an integer multiplication overflows, then the result is the low-order bits of the mathematical product as represented in some sufficiently large two's-complement format. As a result, if overflow occurs, then the sign of the result may not be the same as the sign of the mathematical product of the two operand values.
> ```
> 
> I am interpreting this as using long arithmetic for integer overflow and taking low 32-bits of result. In this sense suggested optimization is legal.
> 
> _But_ we may have bugs in VM. I suggest to add a regression test to compare golden (from Interpreter) values with results from optimized compiled code for all interesting/corner cases.

@vnkozlov Thanks for confirmation.

I added a test as you suggested, but not sure it covers all interesting/corner cases, please advise.

**Review by vnkozlov — COMMENTED** [review]
Repetitive copy-past code pattern concern me - prone to bugs. Can you consider using next pattern instead:
```
   if (op1 == Op_MulI && op2 == Op_MulI) {
     Node* add_in1 = NULL;
     Node* add_in2 = NULL;
     Node* mul_in = NULL;
     if (in1->in(1) == in2->in(1)) {
       // Convert "a*b+a*c into a *(b+c)
       add_in1 = in1->in(2);
       add_in2 = in2->in(2);
       mul_in = in1->in(1);
...
     if (mul_in != NULL) {
       Node* add = phase->transform(new AddINode(add_in1, add_in2));
       return new MulINode(mul_in, add);
     }
```

File: test/hotspot/jtreg/compiler/integerArithmetic/TestIntegerAssociative.java
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/integerArithmetic/TestIntegerAssociative.java:104]: You don't need separate method for compilation. First call will be executed in interpreter.

File: test/hotspot/jtreg/compiler/integerArithmetic/TestIntegerAssociative.java
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/integerArithmetic/TestIntegerAssociative.java:99]: You need tests for all order variations you optimized.

**Review by vnkozlov — COMMENTED** [review]
I am fine with your selection of corner cases in the test.
You need second test file for `long` values.

**Review by vnkozlov — APPROVED** [review]
Good.

You need second review.

**Review by adinn — APPROVED** [review]
This is definitely clearer and the test is a good safety net.

**Comment by zhengyu123** [comment]
@adinn @vnkozlov Thanks for the review

**Pull request closed without merging**
